### PR TITLE
Fix stat metadata and RAM/Redis index parity

### LINKED ISSUE #1008: Redis index store never caches an empty directory listing
## Problem

`RedisIndexCacheStore.set_dir` maps a directory listing onto a Redis list. Redis has no empty list: `RPUSH` with no values creates no key, and `EXPIRE` on a missing key is a no-op. So a listing with zero children is never stored.

Python: `python/mirage/cache/index/redis.py` (`set_dir`, the `if child_keys:` guard). TypeScript has the same shape in `typescript/packages/core/src/cache/index/redis.ts`.

`RAMIndexCacheStore` stores `_children[path] = []` plus an expiry, so it does distinguish "listed, empty" from "never listed". The two stores disagree on the contract.

## Effect

- `ls` on an empty directory is a cold backend call every time.
- Negative knowledge never works under an empty directory: `stat /s3/empty/x` hits the backend on every probe, which is exactly the shell `cd` probing (`.git`, `HEAD`) the index exists to absorb.

## Also: `list_dir` spends three round trips on one fact

`list_dir` runs `EXISTS`, then `TTL`, then `LRANGE`. `TTL == -2` after `EXISTS == 1` cannot happen, Redis drops an expired key before `EXISTS` sees it. Observed with `redis-cli MONITOR` on a warm `ls -l /s3/`:

```
"EXISTS" "index:mirage:idx:children:/s3"
"TTL"    "index:mirage:idx:children:/s3"
"LRANGE" "index:mirage:idx:children:/s3" "0" "-1"
```

Once emptiness is representable, one `LRANGE` (or one `GET`) answers all three states.

## Proposed fix

Store the listing in a form that can be empty, for example a JSON string under the children key, or keep the list and write a sentinel key beside it. Then collapse `list_dir` to a single read.

Add a contract test parametrized over `[ram, redis]` (`tests/cache/index/`) covering: empty listing round-trips as `entries=[]`, missing listing is `NOT_FOUND`, expiry is `EXPIRED`, `invalidate_dir` / `invalidate_prefix`. Today only `test_ram.py` has the empty-listing case, which is why the Redis store drifted. Mirror in TS.

### LINKED ISSUE #1021: [Issue]: TS stat line differs from Python: mergeOverlayStat drops content, S3 modified keeps .000Z
### What type of report is this?

Bug

### Area

TypeScript packages

### Summary

Two places where a TypeScript `stat` line differs from the Python one for the same object.

1. **`mergeOverlayStat` drops `content`.** `typescript/packages/core/src/workspace/mount/namespace/overlay.ts:53-65` ends with a hand-written `new FileStat({...})` that copies eleven fields and omits `content`. Python's `merge_overlay_stat` uses `stat.model_copy(update=update)` and keeps everything. `FileStat.with()` exists for this and its docstring (`types.ts:543-549`) says hand-written lists are how fields go missing. Visible after any `touch` (which writes an mtime into the overlay): `stat` prints `type=file` where Python prints `type=text`.
2. **S3 `modified` keeps `.000Z`.** `core/s3/driver.ts:150` uses `resp.LastModified.toISOString()`; `HeadObject` returns whole seconds, so it renders `2026-09-05T10:55:39.000Z`. Python's `to_iso_z` renders `2026-09-05T10:54:04Z`. The list path (`isoOf`, `driver.ts:55`) has the same gap, plus a digit-count split: Python six fraction digits, TS three, for the same millisecond value, and that one lands in the persisted `remote_time`.

### Steps to reproduce

```
tee /s3/c.txt <<< x ; touch /s3/c.txt ; stat /s3/c.txt
tee /s3/new.txt <<< n ; stat /s3/new.txt
```

### Expected behavior

```
name=c.txt size=2 modified=... type=text
name=new.txt size=2 modified=2026-09-05T10:54:04Z type=text
```

### Actual behavior

```
name=c.txt size=2 modified=... type=file
name=new.txt size=2 modified=2026-09-05T10:55:39.000Z type=text
```

### Mirage version or commit

main @ 67389cca5

### Additional context

Fix for (1) is `return stat.with(update)`. For (2), add a `toIsoZ(d: Date)` beside `epochToIso` in `utils/dates.ts` and use it in `core/s3/driver.ts`; the other bare `toISOString()` sites (`core/s3/watch.ts:34`, `core/databricks_volume/readdir.ts:115`, `node/src/core/ssh/watch.ts:57`) have the same shape. Pick one fraction-digit policy for the list path in both languages.

### LINKED ISSUE #1022: [Issue]: Index store contract differs between RAM and Redis (lost seed, no EXPIRED on Redis, expired_at clamp); no shared contract test
### What type of report is this?

Bug

### Area

Python package, TypeScript packages

### Summary

The two index stores share an interface but not a contract, and there is no test that runs one sequence against both. A live probe against RAM and Redis found these differences beyond #1008:

1. **Redis `seed()` can be lost.** `RedisIndexCacheStore.seed` only stashes the data (`python/mirage/cache/index/redis.py:104-109`); `_flush_seed` runs on the next `get`/`put`/`list_dir`/`set_dir`/`entries`/`invalidate_dir`/`invalidate_prefix`, but not on `clear`, `invalidate` or `close`. `seed()` then `close()` writes nothing. RAM's `seed` is synchronous. Python only: the TS base store (`cache/index/store.ts`) has no `seed()` or `entries()` at all, so the two base interfaces are not parallel either.
2. **`list_dir` never reports `EXPIRED` on Redis after natural TTL expiry.** The key is gone, `EXISTS` says 0, and `NOT_FOUND` is returned; the `TTL == -2` branch (`redis.py:157`) is unreachable. RAM returns `EXPIRED`. The github backend keys its refill on `EXPIRED` (`core/github/readdir.py:46`, `read.py:66`); `ensure_live_index` covers the common case because github seeds the whole tree with one expiry a year out and re-probes the root, but a listing written through `set_dir` with the default TTL (the truncated-tree `_fallback_readdir` path) expires alone and reads as absent. Same in `redis.ts:135-144`.
3. **`set_dir(expired_at <= now)`** is EXPIRED at once on RAM; Redis clamps the TTL to at least 1 s (`redis.py:180-183`, `redis.ts:168-171`) and serves the entries for up to a second.

File cache, by design but worth a comment: Redis has no `cache_size`/`cache_entries`/LRU (external `maxmemory`), and stores `data:` and `meta:` as two keys that server-side eviction can separate, so `is_fresh` can disagree with `get`.

### Steps to reproduce

Probe script: one asyncio sequence over both stores, comparing after every step. Available on request; it should become `tests/cache/index/test_store_contract.py`.

### Expected behavior

Every observable answer identical between RAM and Redis, or the difference declared in one place.

### Actual behavior

See list above; #1008 is the fourth row of the same table.

### Mirage version or commit

main @ 67389cca5

### Additional context

The `invalidate()` docstring in `redis.py` already sketches the right shape for (2): an `invalidated_at` marker compared against `index_time`, with a long safety-net TTL on the key so staleness is decided client-side instead of by Redis deleting the key. Fix (1) by flushing in `clear`/`invalidate`/`close` or by dropping the lazy path to match TS.

### PR DESCRIPTION
**Pull request by zechengz**
Fix Python/TypeScript stat differences and make the RAM and Redis metadata indexes honor the same contract.

`tee /s3/c.txt <<< x; touch /s3/c.txt; stat /s3/c.txt` preserves `type=text`. Stat overlays use `FileStat.with()`, and S3 and Databricks stat timestamps match Python. S3 and SSH watches retain the timestamp spelling used by saved checkpoints.

Redis distinguishes missing, empty, fresh and expired listings while retaining entry metadata. Global and directory invalidation tokens have permanent keys independent of the payload format. One MGET reads a listing and both tokens; scoped invalidation reaches every participating format without expiring unrelated directories. Unique tokens prevent eviction or a refill from reviving old listings. Snapshot seeds accumulate, flush on close, and initialize directory tokens in batches. TypeScript gains the matching seed/entries interface and mount ownership filtering.

Entry and listing prefixes live together in mirrored constants modules and both change with the cache format. The unversioned format is unsupported: current workers neither read nor clean up its payload keys. Workers without the shared-token protocol must be upgraded together; participating workers can roll between payload formats.

Direct lookups validate retained metadata through fresh parent listings. Missing parents force a refresh, including orphaned entries from partial Google listings; newly warmed partial entries can serve the current lookup and are revalidated on the next one. Drive refreshes ancestor IDs, and GitHub/Hugging Face readers replace expired snapshots and remove obsolete entries. Truncated GitHub trees resolve each parent from the current ref; deleted directories report ENOENT and directories replaced by blobs read as files.

Concurrent lookups sharing an index serialize each parent's freshness check, invalidation, warm and final retry. GitHub and Hugging Face readers use a mount-wide guard through snapshot replacement and lookup, including lazy hydration. These guards coordinate tasks within one process and release on errors or cancellation.

Google read/stat/grep and Discord FUSE fixtures now seed complete parent listings, resolving the test failures exposed by parent validation. The TypeScript lockfile upgrades Knip's transitive `smol-toml` from 1.7.0 to 1.8.0, with an override excluding vulnerable versions.

Validation on the updated branch:

- Python main and serial Redis suites: 20,467 passed, 110 skipped, 1 expected failure; camel excluded as in the main CI selection.
- TypeScript core: 804 test files passed; 11,730 passed, 112 skipped, 1 expected failure. Node: 2,973 passed, 302 skipped. Redis-dependent tests skip in these full TypeScript runs.
- Targeted suites with live Redis enabled: 989 Python, 562 TypeScript core, and 129 Node tests passed, covering the cache and affected service consumers.
- Concurrency regressions reproduce the previous implementation's failures and pass with the transaction guards, including expired/missing parents, overlapping partial listings, snapshot replacement, failure recovery and waiter cancellation.
- Shared GitHub/Hugging Face mount and CLI integration batteries: 125 cases passed per language with `--strict`, using the CLI-specific GitHub fixture for permission expectations.
- Core/Node builds and type checks, all applicable pre-commit hooks, and strict layout parity passed.
- The prior update also passed both Redis-index golden examples and `pnpm audit --no-color` with no known vulnerabilities.

Related work reviewed: #1004 preserves Redis entry metadata; #958 adds fresh-read identity behavior; #990 adds file-cache clocks; #1036 reduces S3 metadata requests. Those changes do not resolve these defects.

Fixes #1021.
Fixes #1022.
Fixes #1008.

### CODE REVIEW

File: python/mirage/cache/index/redis.py [hunk 1]
```diff
@@ -249,46 +266,22 @@ async def invalidate_prefix(self, resource_path: str) -> None:
...
-        pre-write tree as if it were current, and quietly wrong is the
-        worse of the two. Closing this properly means an ``invalidated_at``
-        marker key compared against each entry's ``index_time``, which
-        needs no schema change and can ride the same round trip.
-        """
-        await self.clear()
+        await self._flush_seed()
+        # One atomic generation change expires all listings without racing
+        # another client's refill or resurrecting a concurrently removed path.
+        await self._client.set(self._generation_key, uuid7())
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/cache/index/redis.py:334]: Fixed in 889f9ba38. GitHub stat now checks freshness and membership through the parent readdir before using an entry, and file reads share that stat path. This preserves the index contract while preventing stale sizes/SHAs and reads of deleted entries. Both languages have RAM/Redis regressions for edits and deletions after invalidation, including truncated trees; repeated fresh lookups make no further tree requests. The full Python and TypeScript core suites and pre-commit checks pass.

File: python/mirage/core/github/readdir.py [hunk 2]
```diff
@@ -48,10 +48,11 @@ async def readdir(
             listing = await index.list_dir(virtual_key)
     if listing.entries is not None:
         return listing.entries
+    if accessor.truncated and listing.status in (LookupStatus.NOT_FOUND,
+                                                 LookupStatus.EXPIRED):
+        return await _fallback_readdir(accessor, virtual_key, index, virtual,
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/core/github/readdir.py:66]: Fixed in 889f9ba38. The truncated-tree fallback now resolves from the current ref through each parent tree, without trusting cached directory SHAs. It also rejects path segments that are now files or missing. Paired RAM/Redis regressions keep the parent listings fresh while expiring a nested directory, advance both ancestor SHAs, and verify that the new contents are cached and reused. Python and TypeScript tests pass.

File: python/mirage/core/github/readdir.py [hunk 3]
```diff
@@ -129,7 +128,7 @@ async def _resolve_dir_sha(
                                        accessor.pool)
         found = False
         for entry in entries:
-            if entry.path == part:
+            if entry.path == part and entry.type == "tree":
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/core/github/readdir.py]: Fixed in eb3ebc678. Both implementations now cache every complete parent listing fetched while resolving a truncated-tree path, and remove a former directory subtree before caching a replacement blob. RAM/Redis regressions expire the nested directory while keeping its parent fresh, then verify that stat/read report ENOENT after deletion or return the replacement file and its new SHA. The shared GitHub integration target passes all 23 cases in each language.

File: python/mirage/cache/index/redis.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/cache/index/redis.py:334]: Fixed in eb3ebc678. After a successful tree fetch, GitHub now removes the previous snapshot under this mount's prefix before seeding the new tree. Deleted directories and directories replaced by files no longer retain expired listings or trigger repeated recursive fetches. Paired RAM/Redis tests verify ENOENT on repeated readdir calls, exactly one refill, removal of old descendant metadata, and preservation of entries outside the mount.

[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/cache/index/redis.py:334]: Fixed in eb3ebc678. The shared entry_or_warm/entryOrWarm helper now validates parent membership and refreshes expired listings before returning retained metadata, clearing old children before a partial refresh. Drive's directory-ID resolution uses the same helper so renamed/replaced ancestors cannot leave stale IDs usable. Hugging Face checks root/parent listing freshness and replaces the old snapshot on refill. RAM/Redis regressions cover deleted files, changed IDs/metadata, renamed Drive folders, removed Hugging Face directories, and refresh failures. Python passed 18,902 tests across the broad and serial CLI runs; TypeScript core passed 11,589 tests, including live Redis coverage.

File: typescript/packages/core/src/core/s3/watch.ts [hunk 4]
```diff
@@ -31,7 +32,7 @@ interface ListedObject {
 
 function isoOf(value: Date | string | undefined): string | null {
   if (value === undefined) return null
-  return value instanceof Date ? value.toISOString() : value
+  return value instanceof Date ? toIsoZ(value) : value
```
[the remark this answers is not in the document]
Comment (zechengz) [reply typescript/packages/core/src/core/s3/watch.ts]: Fixed in eb3ebc678. S3 and SSH watchers retain the previous toISOString() spelling used by persisted fingerprints; the stat timestamp normalization remains separate. New checkpoint regressions in both watchers start with literal pre-upgrade fingerprints at whole-second and fractional-second timestamps. Unchanged pulls emit no events and preserve the checkpoint; a real mtime change still emits UPDATE.

File: typescript/packages/core/src/cache/index/redis.ts [hunk 5]
```diff
@@ -199,45 +280,39 @@ export class RedisIndexCacheStore extends IndexCacheStore {
...
-  // the pre-write tree as if it were current, and quietly wrong is the worse
-  // of the two. Closing this properly means an `invalidatedAt` marker key
-  // compared against each entry's indexTime, which needs no schema change and
-  // can ride the same round trip.
   async invalidate(): Promise<void> {
-    await this.clear()
+    await this.flushSeed()
+    const c = await this.client()
+    // Atomically expire listings without overwriting concurrent refills/deletions.
+    await c.set(this.generationKey, uuid7())
```
[the remark this answers is not in the document]
Comment (zechengz) [reply typescript/packages/core/src/cache/index/redis.ts:336]: Fixed in 57fd6757f. The Node Hugging Face implementation now mirrors Python: expired roots trigger refills, direct entry lookups check parent listing freshness, and a successful fetch replaces the snapshot under the mount prefix so deleted entries and directories are removed. Cleanup happens only after a successful fetch, and the root lookup never consults a parent outside the mount. Added 30 RAM/Redis regression cases for direct stat/read after edits or deletions, replaced directories, independently expired parents, retry after fetch failure, and mount isolation; 28 reproduced the defect before the fix. Validation: 185 HF backend/resource/CLI tests, 2,927 full Node tests (4 skipped), and all 43 shared HF mount/CLI cases in each language pass. Node build, type checking, lint, and layout parity also pass.

File: python/mirage/cache/index/redis.py [hunk 6]
```diff
@@ -12,24 +12,24 @@
...
+from mirage.cache.index.config import (IndexDirectory, IndexEntry, ListResult,
+                                       LookupResult, LookupStatus)
 from mirage.cache.index.store import IndexCacheStore
 from mirage.core.timeutil import to_iso_z
+from mirage.utils.ids import uuid7
 from mirage.utils.key_prefix import under_path
 
 ENTRY_PREFIX = "mirage:idx:entry:"
-CHILDREN_PREFIX = "mirage:idx:children:"
+CHILDREN_PREFIX = "mirage:idx:directory:v2:"
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/cache/index/redis.py]: Fixed in 03dd396ce. Both Python and TypeScript now store entry rows under mirage:idx:entry:v2:, matching the versioned directory records. Legacy rows are excluded from get() and entries(), so an upgraded cache starts cold and direct service lookups warm from the backend. Added six upgrade regressions per language covering updates, renames and deletions, with and without invalidate(); all 12 failed before the fix and pass now. Validation: 145 Python index tests, 107 TypeScript index/reconciliation tests, the full TypeScript core suite (11,606 passed, 1 expected failure), Redis-backed examples, type checks and pre-commit checks pass.

File: python/mirage/cache/index/warm.py [hunk 7]
```diff
@@ -47,14 +47,26 @@ async def entry_or_warm(
...
             index; ``None`` when the key has no distinct parent to list.
     """
+    parent = virtual_key.rstrip("/").rsplit("/", 1)[0] or "/"
+    listing = await index.list_dir(parent)
+    if listing.entries is not None and virtual_key not in listing.entries:
+        return None
     hit = await index.get(virtual_key)
-    if hit.entry is not None:
+    if hit.entry is not None and listing.status != LookupStatus.EXPIRED:
         return hit.entry
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/cache/index/warm.py]: Fixed in e9e5e7033. `entry_or_warm`/`entryOrWarm` now accepts a retained entry only when a fresh parent listing includes it. Missing or expired parents invalidate the target before warming, so partial `put()` rows cannot survive a GWS update, delete, or rename; a newly warmed put-only row is usable for that call and is revalidated on the next lookup. Added RAM, fake/live Redis, GDocs, and direct GDrive regressions in both languages, and updated cache-hit fixtures to seed complete parent listings.

File: python/mirage/core/github/tree.py [hunk 8]
```diff
@@ -213,6 +218,8 @@ async def refill_index(
     accessor.truncated = truncated
     accessor.tree = tree
     accessor.tree_loaded = True
+    # A refill replaces this mount's snapshot, including paths now absent.
+    await index.invalidate_prefix(prefix.rstrip("/") or "/")
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/core/github/tree.py:224]: Fixed in 416fffe0f. GitHub and Hugging Face now guard each mount's snapshot readers from the freshness check through replacement and the final entry/listing lookup. Concurrent readers sharing an index wait, then recheck freshness, so only one refills an expired snapshot. The same guard covers index-backed lazy hydration. Python and TypeScript regressions exercise parallel stat/listing calls over RAM and Redis and assert current metadata plus a single fetch. The affected live-Redis suites and all 125 shared GitHub/Hugging Face mount/CLI cases per language pass. These guards coordinate callers sharing an index in one process.

File: python/mirage/cache/index/warm.py [hunk 9]
```diff
@@ -47,14 +52,30 @@ async def entry_or_warm(
...
     hit = await index.get(virtual_key)
-    if hit.entry is not None:
+    if hit.entry is not None and listing.entries is not None:
         return hit.entry
     if warm is None:
         return None
+    if listing.status == LookupStatus.EXPIRED:
+        # Retained metadata is not proof of existence. Drop the old children
+        # before warming, including when a best-effort listing only puts rows.
+        await index.invalidate_dir(parent)
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/cache/index/warm.py]: Fixed in 416fffe0f. Both entry_or_warm/entryOrWarm implementations now serialize the complete parent freshness check, invalidation, warm and retry under a shared per-index/per-parent guard. Queued lookups recheck after acquiring it, while a partial warm retains its own row through the final lookup. Added regressions for simultaneous siblings under expired/missing parents, overlapping partial refreshes, failure recovery and Python waiter cancellation. The previous implementation fails the concurrency regressions; the updated Python suite passes 20,467 tests, and the full TypeScript core/Node suites, targeted live Redis tests, builds, type checks, lint and layout parity pass.
